lib/validator.ex

```diff
@@ -70,24 +103,26 @@ defmodule UltraDark.Validator do
     |> Enum.all?(&(&1 == true))
   end
 
-  @spec valid_transactions?(Block) :: :ok | {:error, String.t}
+  @spec valid_transactions?(Block) :: :ok | {:error, :invalid_inputs}
   def valid_transactions?(%{transactions: transactions}) do
-    if Enum.all?(transactions, &(valid_transaction? &1)), do: :ok, else: {:error, "Transaction contains invalid input(s)"}
+    if Enum.all?(transactions, &valid_transaction?(&1)), do: :ok, else: {:error, :invalid_inputs}
   end
 
+  @spec is_coinbase?(Transaction) :: :ok | {:error, {:not_coinbase, String.t()}}
   defp is_coinbase?(tx) do
-    if tx.txtype == "COINBASE", do: :ok, else: {:error, "Transaction 1 of block is not of txtype COINBASE"}
+    if tx.txtype == "COINBASE", do: :ok, else: {:error, {:not_coinbase, tx.txtype}}
   end
 
+  @spec appropriate_coinbase_output?(list, number) :: :ok | {:error, :invalid_coinbase}
   defp appropriate_coinbase_output?([coinbase | transactions], block_index) do
     total_fees = Block.total_block_fees(transactions)
     reward = Block.calculate_block_reward(block_index)
     amount = List.first(coinbase.outputs).amount
-    if D.equal?(D.add(total_fees, reward), amount), do: :ok, else: {:error, "Coinbase output is invalid"}
+    if D.equal?(D.add(total_fees, reward), amount), do: :ok, else: {:error, :invalid_coinbase}
```

Comment:
This should correspond directly to what makes the coinbase invalid. Saying that the coinbase is invalid isn't specific enough, as we don't know _why_ the coinbase is invalid by looking at the error message